Sources/FirebaseFirestore/Query+Swift.swift

```diff
@@ -39,4 +39,147 @@ extension Query {
       })
     }
   }
+
+  public func addSnapshotListener(_ listener: @escaping (QuerySnapshot?, Error?) -> Void) -> ListenerRegistration {
+    addSnapshotListener(includeMetadataChanges: false, listener: listener)
+  }
+
+  public func addSnapshotListener(includeMetadataChanges: Bool, listener: @escaping (QuerySnapshot?, Error?) -> Void) -> ListenerRegistration {
+    typealias ListenerCallback = (QuerySnapshot?, Error?) -> Void
+    let boxed = Unmanaged.passRetained(listener as AnyObject)
+    let instance = swift_firebase.swift_cxx_shims.firebase.firestore.query_add_snapshot_listener(self, { snapshot, errorCode, errorMessage, pvListener in
+      let callback = Unmanaged<AnyObject>.fromOpaque(pvListener!).takeUnretainedValue() as! ListenerCallback
+
+      let error = NSError.firestore(errorCode, errorMessage: errorMessage)
+      // We only return a snapshot if the error code isn't 0 (aka the 'ok' error code)
+      let returned = error == nil ? snapshot?.pointee : nil
+
+      // Make sure we dispatch our callback back into the main thread to keep consistent
+      // with the reference API which will call back on the 'user_executor' which typically
+      // ends up being the main queue.
+      // Relevant code:
+      // - https://github.com/firebase/firebase-ios-sdk/blob/main/Firestore/Source/API/FIRFirestore.mm#L210-L218
+      // - https://github.com/firebase/firebase-ios-sdk/blob/main/Firestore/core/src/api/document_reference.cc#L236-L237
+      DispatchQueue.main.async {
+        callback(returned, error)
+      }
```

Comment:
This was specifically for `DocumentReference` do we know if the same assumptions about which return queue is used can hold true for Query?

Reply:
I read the code and tested it. The callbacks in the Obj C API are intended to run on the main queue, but can be overridden thru FirebaseSettings.